lib/posix/pthread.c

```diff
@@ -151,69 +286,85 @@ static void zephyr_thread_wrapper(void *arg1, void *arg2, void *arg3)
  *
  * See IEEE 1003.1
  */
-int pthread_create(pthread_t *newthread, const pthread_attr_t *_attr,
-		   void *(*threadroutine)(void *), void *arg)
+int pthread_create(pthread_t *th, const pthread_attr_t *_attr, void *(*threadroutine)(void *),
+		   void *arg)
 {
-	int rv;
-	int32_t prio;
+	int err;
 	k_spinlock_key_t key;
-	uint32_t pthread_num;
-	k_spinlock_key_t cancel_key;
-	pthread_condattr_t cond_attr;
-	struct posix_thread *thread;
-	const struct pthread_attr *attr = (const struct pthread_attr *)_attr;
+	struct posix_thread *t;
+	pthread_barrier_t barrier;
+	struct pthread_attr attr;
 
-	/*
-	 * FIXME: Pthread attribute must be non-null and it provides stack
-	 * pointer and stack size. So even though POSIX 1003.1 spec accepts
-	 * attrib as NULL but zephyr needs it initialized with valid stack.
-	 */
-	if ((attr == NULL) || (attr->initialized == 0U)
-	    || (attr->stack == NULL) || (attr->stacksize == 0)) {
+	if (!pthread_attr_is_valid((const struct pthread_attr *)_attr)) {
 		return EINVAL;
 	}
 
-	key = k_spin_lock(&pthread_pool_lock);
-	for (pthread_num = 0;
-	    pthread_num < CONFIG_MAX_PTHREAD_COUNT; pthread_num++) {
-		thread = &posix_thread_pool[pthread_num];
-		if (thread->state == PTHREAD_EXITED || thread->state == PTHREAD_TERMINATED) {
-			thread->state = PTHREAD_JOINABLE;
-			break;
+	/* copy the attribute in to the */
+	(void)memcpy(&attr, _attr, sizeof(attr));
+
+	key = k_spin_lock(&q_lock);
+	/* recycle old nodes if they are available */
+	do {
+		struct posix_thread *safe_t;
+
+		SYS_DLIST_FOR_EACH_CONTAINER_SAFE(&done_q, t, safe_t, q_node) {
+			if (t->detachstate == PTHREAD_CREATE_JOINABLE) {
```

Comment:
Could this transition be managed along with the manipulation of detachstate and/or moving objects into done_q? It seems a bit odd to have threads hanging in limbo until someone goes to create another thread.

Reply:
yeah, workqueue def seems like overkill. Was wondering if it could be done when the state transition is made though; you're just looking for threads in the done_q which are JOINABLE, so that should be detectable when either adding to done_q or switching to JOINABLE, it seems. Might be a future cleanup though; this solution looks correct.